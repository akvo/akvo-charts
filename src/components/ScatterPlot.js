import React, {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useRef
} from 'react';
import * as echarts from 'echarts';
import transformConfig from '../utils/transformConfig';

import styles from '../styles.module.css';

const getOptions = ({ data = [], symbolSize = 20 }) => {
  return {
    series: [
      {
        symbolSize,
        data: data.filter((d) => Array.isArray(d)),
        type: 'scatter'
      }
    ]
  };
};

const ScatterPlot = forwardRef(
  ({ config, data, symbolSize = 20, isTest = false }, ref) => {
    const chartRef = useRef(null);

    useImperativeHandle(ref, () => ({
      getChartInstance: () => {
        return chartRef.current
          ? echarts.getInstanceByDom(chartRef.current)
          : null;
      }
    }));

    useEffect(() => {
      let chart;

      if (chartRef.current) {
        chart = isTest
          ? echarts.init(chartRef.current, 'light', {
              renderer: 'svg',
              width: 400,
              height: 400
            })
          : echarts.init(chartRef.current, 'light', { renderer: 'svg' });
        const options = {
          ...transformConfig({ ...config, symbolSize }),
          ...getOptions({ data })
        };
        chart.setOption(options);
      }

      return () => {
        if (chart) {
          chart.dispose();
        }
      };
    }, [config, data, isTest, symbolSize]);

    return (
      <div
        ref={chartRef}
        role="figure"
        className={styles.container}
        data-testid="scatter-plot"
      />
    );
  }
);

export default ScatterPlot;
