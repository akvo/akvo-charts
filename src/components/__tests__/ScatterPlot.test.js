import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import ScatterPlot from '../ScatterPlot';

describe('ScatterPlot chart', () => {
  test('renders ScatterPlot component with 2d array data format', () => {
    const input2D = [
      ['A', 1, 2],
      ['B', 1, 3],
      ['C', 5, 7]
    ];

    const config = {
      title: 'ScatterPlot Chart Example',
      renderer: 'svg',
      width: 400,
      height: 400
    };

    render(
      <ScatterPlot
        config={config}
        data={input2D}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders ScatterPlot component with row based key-value format (object array)', () => {
    const inputKeyValue = [
      { label: 'A', x: 1, y: 2 },
      { label: 'B', x: 1, y: 3 },
      { label: 'C', x: 5, y: 7 }
    ];
    const config = {
      title: 'ScatterPlot Chart Example',
      renderer: 'svg',
      width: 400,
      height: 400
    };

    render(
      <ScatterPlot
        config={config}
        data={inputKeyValue}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('renders ScatterPlot component with column based key-value format', () => {
    const inputObj = {
      A: [1, 2],
      B: [1, 3],
      C: [5, 7]
    };

    const config = {
      title: 'ScatterPlot Chart Example',
      renderer: 'svg',
      width: 400,
      height: 400
    };

    render(
      <ScatterPlot
        config={config}
        data={inputObj}
        symbolSize={50}
      />
    );

    const chartContainer = screen.getByRole('figure');
    expect(chartContainer).toBeInTheDocument();
  });

  test('matches ScatterPlot snapshot', async () => {
    const data = [
      ['A', 1, 2],
      ['B', 1, 3],
      ['C', 5, 7]
    ];

    const config = {
      title: 'ScatterPlot Chart Example',
      renderer: 'svg',
      width: 400,
      height: 400
    };

    const ref = React.createRef();
    render(
      <ScatterPlot
        config={config}
        data={data}
        ref={ref}
      />
    );

    await waitFor(() => {
      expect(ref.current.renderToSVGString()).toMatchSnapshot();
    });
  });
});
