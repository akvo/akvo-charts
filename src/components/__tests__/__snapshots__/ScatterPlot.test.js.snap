// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScatterPlot matches ScatterPlot snapshot 1`] = `
<div>
  <div
    _echarts_instance_="ec_1722395092265"
    class="container"
    data-testid="scatter-plot"
    role="figure"
    style="position: relative;"
  >
    <div
      style="position: relative; overflow: hidden; width: 400px; height: 400px;"
    >
      <svg
        baseProfile="full"
        height="400"
        style="position: absolute; left: 0px; top: 0px; user-select: none;"
        version="1.1"
        width="400"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <rect
          fill="none"
          height="400"
          width="400"
          x="0"
          y="0"
        />
        <g>
          <path
            d="M37.44 340.5L384 340.5"
            fill="transparent"
            stroke="#E0E6F1"
          />
          <path
            d="M37.44 292.5L384 292.5"
            fill="transparent"
            stroke="#E0E6F1"
          />
          <path
            d="M37.44 244.5L384 244.5"
            fill="transparent"
            stroke="#E0E6F1"
          />
          <path
            d="M37.44 196.5L384 196.5"
            fill="transparent"
            stroke="#E0E6F1"
          />
          <path
            d="M37.44 148.5L384 148.5"
            fill="transparent"
            stroke="#E0E6F1"
          />
          <path
            d="M37.44 100.5L384 100.5"
            fill="transparent"
            stroke="#E0E6F1"
          />
          <path
            d="M37.44 340.5L384 340.5"
            fill="transparent"
            stroke="#000"
            stroke-linecap="round"
          />
          <path
            d="M37.5 340L37.5 345"
            fill="transparent"
            stroke="#000"
          />
          <path
            d="M124.5 340L124.5 345"
            fill="transparent"
            stroke="#000"
          />
          <path
            d="M210.5 340L210.5 345"
            fill="transparent"
            stroke="#000"
          />
          <path
            d="M297.5 340L297.5 345"
            fill="transparent"
            stroke="#000"
          />
          <path
            d="M384.5 340L384.5 345"
            fill="transparent"
            stroke="#000"
          />
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="end"
            transform="translate(29.44 340)"
          >
            0
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="end"
            transform="translate(29.44 292)"
          >
            2
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="end"
            transform="translate(29.44 244)"
          >
            4
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="end"
            transform="translate(29.44 196)"
          >
            6
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="end"
            transform="translate(29.44 148)"
          >
            8
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="end"
            transform="translate(29.44 100)"
          >
            10
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="middle"
            transform="translate(80.76 348)"
            y="6"
          >
            10
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="middle"
            transform="translate(167.4 348)"
            y="6"
          >
            8.07
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="middle"
            transform="translate(254.04 348)"
            y="6"
          >
            13
          </text>
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 12px; font-family: sans-serif;"
            text-anchor="middle"
            transform="translate(340.68 348)"
            y="6"
          >
            9.05
          </text>
          <path
            d="M1 0A1 1 0 1 1 1 -0.0001"
            fill="#4475B4"
            fill-opacity="0"
            transform="matrix(0,0,0,0,80.76,147.04)"
          />
          <path
            d="M1 0A1 1 0 1 1 1 -0.0001"
            fill="#4475B4"
            fill-opacity="0"
            transform="matrix(0,0,0,0,167.4,173.2)"
          />
          <path
            d="M1 0A1 1 0 1 1 1 -0.0001"
            fill="#4475B4"
            fill-opacity="0"
            transform="matrix(0,0,0,0,254.04,158.08)"
          />
          <path
            d="M1 0A1 1 0 1 1 1 -0.0001"
            fill="#4475B4"
            fill-opacity="0"
            transform="matrix(0,0,0,0,340.68,128.56)"
          />
          <path
            d="M-5 -5l10 0l0 10l-10 0Z"
            fill="rgb(0,0,0)"
            fill-opacity="0"
            stroke="#ccc"
            stroke-width="0"
            transform="translate(200 40)"
          />
          <path
            d="M-87.74 -5l175.48 0l0 24l-175.48 0Z"
            fill="rgb(0,0,0)"
            fill-opacity="0"
            stroke="#ccc"
            stroke-width="0"
            transform="translate(205 5)"
          />
          <text
            dominant-baseline="central"
            fill="#000"
            style="font-size: 14px; font-family: sans-serif; font-weight: bold;"
            text-anchor="middle"
            transform="translate(205 5)"
            xml:space="preserve"
            y="7"
          >
            ScatterPlot Chart Example
          </text>
        </g>
      </svg>
    </div>
    <div
      class=""
    />
  </div>
</div>
`;
